### PR TITLE
Feature mtc 147 migrate repos to sql

### PR DESCRIPTION
**Pull request by GSailen**

### CODE REVIEW

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountBalanceChangeEntity.cs [hunk 1]
```diff
@@ -2,77 +2,46 @@
...
-        /// Client id
-        /// </summary>
         public string ClientId { get; set; }
 
-        /// <summary>
-        /// The balance diff
-        /// </summary>
-        public decimal ChangeAmount { get; set; }
+        decimal IAccountBalanceChange.ChangeAmount => (decimal) ChangeAmount;
+        public double ChangeAmount { get; set; }
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountBalanceChangeEntity.cs]: AzureTableEntity supports decimals and complex types (hope so). No need to do like this.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountEntity.cs [hunk 2]
```diff
@@ -59,7 +61,9 @@ public bool IsDisabled
                 MarkValueTypePropertyAsDirty(nameof(IsDisabled));
             }
         }
-        
+
+        public DateTimeOffset ModificationTimestamp { get; set; }
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountEntity.cs]: Why DateTimeOffset?

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangeEntity.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        public string AccountId { get; set; }
+        
+        public DateTime ChangeTimestamp { get; set; }
+        
+        public string Id { get; set; }
+
+        public string ClientId { get; set; }
+
+        decimal IAccountBalanceChange.ChangeAmount => (decimal) ChangeAmount;
+        public double ChangeAmount { get; set; }
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangeEntity.cs]: For SQL 100% no need to do like this.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                    _log?.WriteErrorAsync(nameof(AccountBalanceChangesRepository), "CreateTableIfDoesntExists", null, ex);
+                    throw;
+                }
+            }
+        }
+        
+        public async Task<List<AccountBalanceChange>> GetAsync(string[] accountIds, DateTime? @from, DateTime? to)
+        {
+            var whereClause = "WHERE " + (accountIds.Any() ? $"AccountId IN ('{string.Join("','", accountIds)}')" : "")
+                                       + (from != null ? " AND ChangeTimestamp > @from" : "")
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs]: If !accountIds.Any() will be wrong clause.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                                       + (to != null ? " AND ChangeTimestamp < @to" : "");
+            
+            using (var conn = new SqlConnection(_settings.Db.SqlConnectionString))
+            {
+                var data = await conn.QueryAsync<AccountBalanceChangeEntity>(
+                    $"SELECT * FROM {TableName} {whereClause}", new { from, to });
+
+                return data.Select(x => _convertService.Convert<AccountBalanceChange>(x)).ToList();
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs]: Why you added interface IAccountBalanceChange and do not use it? Return type could be List<IAccountBalanceChange> or even IEnumerable<IAccountBalanceChange> and additional convertation will not be needed.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            }
+        }
+
+        public async Task<Account> GetAsync(string clientId, string accountId)
+        {
+            using (var conn = new SqlConnection(_settings.Db.SqlConnectionString))
+            {
+                var whereClause = (string.IsNullOrWhiteSpace(clientId) && string.IsNullOrWhiteSpace(accountId) 
+                    ? ""
+                    : "WHERE 1=1 ")
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs]: ???

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        {
+            return await GetAccountAndUpdate(accountId, account => { account.TradingConditionId = tradingConditionId; });
+        }
+
+        public async Task<Account> ChangeIsDisabledAsync(string clientId, string accountId, bool isDisabled)
+        {
+            return await GetAccountAndUpdate(accountId, account => { account.IsDisabled = isDisabled; });
+        }
+
+        private async Task<Account> GetAccountAndUpdate(string accountId, Action<AccountEntity> handler)
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs]: What about concurrency? I think we need to add version property to account and use it in update operation. Should be discussed.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                    {
+                        var entity = Convert(factory());
+
+                        await conn.ExecuteAsync(
+                            $"insert into {TableName} ({GetColumns}) values ({GetFields})", entity);
+
+                        return Convert<TData>(entity);
+                    }
+
+                    return operationInfo.Select(Convert<TData>).FirstOrDefault();
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs]: And what if we will have more than 1? What does it mean? It is error and should not be the case? If yes, better to use Single()

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public async Task<OperationExecutionInfo<TData>> GetAsync<TData>(string operationName, string id) where TData : class
+        {
+            using (var conn = new SqlConnection(_settings.Db.SqlConnectionString))
+            {
+                var operationInfo = await conn.QueryAsync<OperationExecutionInfoEntity>(
+                    $"SELECT * FROM {TableName} WHERE Id = @id", new { id });
+
+                return operationInfo == null || !operationInfo.Any() ? null : Convert<TData>(operationInfo.First());
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs]: Again First(), but looks like should be Single()

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoEntity.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿using Lykke.AzureStorage.Tables;
+using Lykke.AzureStorage.Tables.Entity.Annotation;
+using MarginTrading.AccountsManagement.InternalModels.Interfaces;
+using Newtonsoft.Json;
+
+namespace MarginTrading.AccountsManagement.Repositories.Implementation.SQL
+{
+    public class OperationExecutionInfoEntity : IOperationExecutionInfo<object>
+    {
+        public string Version { get; set; }
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoEntity.cs]: Where Version is used?

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountEntity.cs (hunk 2 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountEntity.cs]: Have no idea - made by Mikhail, will change to DateTime.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangeEntity.cs (hunk 3 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangeEntity.cs]: Interface property backing removed.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountBalanceChangeEntity.cs (hunk 1 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/AzureStorage/AccountBalanceChangeEntity.cs]: Removed.

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs (hunk 4 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs]: Right! Added 1=1 ;)

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs (hunk 6 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs]: Greatest feature of SQL :))

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs (hunk 7 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs]: IsolationLevel.Serializable used

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs (hunk 8 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs]: Changed to QueryFirstOrDefaultAsync

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs (hunk 9 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs]: Changed to QuerySingleOrDefaultAsync

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoEntity.cs (hunk 10 above)
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoEntity.cs]: Not used, removed

**Pull request merged**